kong/resty/dns/client.lua

```diff
@@ -702,6 +702,14 @@ local function parseAnswer(qname, qtype, answers, try_list)
   return true
 end
 
+local function close_socks(resolver)
```

Comment:
do we have an upstream PR for this? if not can we send one (for `lua-resty-dns`)

Reply:
hmmm, then maybe we should look for the root-cause first, before fixing it.

are we keeping them cached somewhere? such that they stay alive instead of getting GC'ed?

Reply:
also; an upstream PR for always closing the sockets would be a sanity item imo. So would still make sesne to me.

Reply:
seems not being cached somewhere, the resolver is created in `ngx.timer`, I'm not sure if the timer matters. However the logic of dns queries is similar between in 2.x and 3.x, so I have no idea why it just happens in 3.x.

Reply:
@ADD-SP and I did further investigations and found that this is caused by timer-ng lib. After replacing the timer-ng with the native timer to schedule the dns query, this issue never happens. I'm not sure but it seems like there are some differences between timer-ng and native time over garbage collection.

Even if the issue is caused by timer-ng, it still makes sense to me that the sockets should be released explicitly.

Reply:
This is very strange, we try to make timer-ng run full GC after each handler and this issue does not appear.

So there is no leak, but it looks like the GC has slowed down.

Reply:
@Tieske I committed an upstream PR: https://github.com/openresty/lua-resty-dns/pull/64
Can we just take this PR as a workaround so that this issue could be fixed quickly? If the PR for resty-dns is released, we can revert this PR and make a better one.

Reply:
Based on OpenResty's claim at [here](https://github.com/openresty/lua-nginx-module#udpsockclose), there are three ways to release socket objects:

1. Call the `close()` method explicitly.
2. Released by the Lua GC.
3. Current client HTTP request finishes processing.

Since lua-resty-dns does not explicitly close socket objects or expose the `close()` method to users, the relevant socket objects can only be released through methods 2 and 3. 

The reason why the native timer is not problematic is that once the timer ends, the fake request is considered finished (effectively fulfilling the condition of method 3). However, timer_ng reuses the existing ngx_timer to execute tasks, so the expected condition of method 3 did not occur and it was only dependent on method 2. I believe this is why those UDP socket objects were leaked.

Reply:
> Can we just take this PR as a workaround so that this issue could be fixed quickly? If the PR for resty-dns is released, we can revert this PR and make a better one.

Yes, that'd work. Please create an item to track this. eg a PR based on the proposed changes in resty-dns, or an internal Jira ticket, whatever works for you.